kv/dist_sender.go

```diff
@@ -556,6 +556,41 @@ func (ds *DistSender) Send(_ context.Context, call client.Call) {
 
 			if err != nil {
 				log.Warningf("failed to invoke %s: %s", call.Method(), err)
+
+				// TODO(tschottdorf): For certain errors we want to expire
+				// the leader and/or range metadata cache. Two scenarios
+				// are relevant:
+				// - a node dies; in that case we do not want to wait until
+				//   we contact someone else to get our NotLeaderError.
+				// - we have a stale range descriptor cached that does not
+				//   contain all replicas, and the ones it contains are down.
+				//   Sounds rare, but if you fire up a test cluster and kill
+				//   the leader that's exactly where you are going to be,
+				//   and that would result in deadlock.
+				//
+				// Unfortunately, the RPC errors are marked retryable, so
+				// for the time being we'll aggressively clear the caches.
+				// This is absolutely not a good idea for most production
+				// environments, but should be helpful on our way to
+				// stabilizing the system. The DistSender really needs to
+				// know on a per-error basis whether it ought to clear
```

Comment:
If I'm understanding this correctly, we might be able to make this distinction based on whether the error was returned by sendRPC() or it came from reply.Header().GoError(). Errors from sendRPC should always invalidate the cache, while errors from Header().GoError() only do if they are of one of the types below.